Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs

```diff
@@ -161,24 +164,46 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
             request.Headers.Authorization = cachedAuthentication;
         }
 
-        var response = await base.SendAsync(request, cancellationToken);
-        if (response is { StatusCode: HttpStatusCode.OK })
-        {
-            return response;
-        }
-        else if (response is { StatusCode: HttpStatusCode.Unauthorized } && TryParseAuthenticationInfo(response, out string? scheme, out AuthInfo? authInfo))
+        int retryCount = 0;
+
+        while (retryCount < MaxRequestRetries)
         {
-            if (await GetAuthenticationAsync(scheme, authInfo.Realm, authInfo.Service, authInfo.Scope, cancellationToken) is AuthenticationHeaderValue authentication)
+            try
             {
-                request.Headers.Authorization = AuthHeaderCache.AddOrUpdate(request.RequestUri, authentication);
-                return await base.SendAsync(request, cancellationToken);
+                var response = await base.SendAsync(request, cancellationToken);
+                if (response is { StatusCode: HttpStatusCode.OK })
+                {
+                    return response;
+                }
+                else if (response is { StatusCode: HttpStatusCode.Unauthorized } && TryParseAuthenticationInfo(response, out string? scheme, out AuthInfo? authInfo))
+                {
+                    if (await GetAuthenticationAsync(scheme, authInfo.Realm, authInfo.Service, authInfo.Scope, cancellationToken) is AuthenticationHeaderValue authentication)
+                    {
+                        request.Headers.Authorization = AuthHeaderCache.AddOrUpdate(request.RequestUri, authentication);
+                        return await base.SendAsync(request, cancellationToken);
+                    }
+                    return response;
+                }
+                else
+                {
+                    return response;
+                }
+            }
+            catch (HttpRequestException e) when (e.InnerException is IOException ioe && ioe.InnerException is SocketException se)
+            {
+                retryCount += 1;
+
+                // TODO: log in a way that is MSBuild-friendly
+                Console.WriteLine($"Encountered a SocketException with message \"{se.Message}\". Pausing before retry.");
+
+                await Task.Delay(TimeSpan.FromSeconds(1.0 * Math.Pow(2, retryCount)));
```

Comment:
nit: can you flow the cancellationtoken into this call? 10+ seconds is a long time to wait and some users may jump the gun to cancel the operation.

Reply:
We do not plumb the token through well today but did this so when we do it Just Works.